flocker/common/algebraic.py

```diff
@@ -0,0 +1,49 @@
+# -*- test-case-name: flocker.common.test.test_algebraic -*-
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+Invariants for algebraic data types.
+"""
+
+from pyrsistent import PClass, field
+
+__all__ = ["StateInvariant"]
+
+
+class StateInvariant(PClass):
+    """
+    An invariant that ensure the given object has a ``state`` attribute
+    in the given states, and that all the other specified attributes are
+    present if and only if the object is in one of the corresponding states.
+
+    :param set allowed_states: Set of allowed states.
+    :param dict expected_attributes: Dictionary mapping attribute names to
+        the set of states that attribute should be present in.
```

Comment:
As interim steps, I suggest the following:

* Make `allowed_states` and `expected_attributes` both private
* Add helpers: `get_states_for_attribute(self, attribute)`, `get_attributes_for_state(self, state)` (or "tag" or "iter_" or whatever)
* Write the tests and possibly `__call__` in terms of those helpers
* See what's remaining in terms of access to private access and proceed accordingly

At the very least, I think a helper to get valid attributes for a given state would simplify a good chunk of code.